File: tests/python/test_timers.py
"""Timers: run a callback later, or on a repeat."""
import loom2d as loom


def test_after_fires_once_when_the_delay_elapses():
    timers = loom.Timers()
    calls = []
    timers.after(1.0, lambda: calls.append("boom"))

    timers.update(0.5)
    assert calls == []
    timers.update(0.5)
    assert calls == ["boom"]

    timers.update(10.0)
    assert calls == ["boom"]          # and never again
    assert timers.count == 0


def test_every_repeats_until_cancelled():
    timers = loom.Timers()
    ticks = []
    handle = timers.every(0.25, lambda: ticks.append(1))

    for _ in range(4):
        timers.update(0.25)
    assert len(ticks) == 4
    assert timers.active(handle)

    assert timers.cancel(handle)
    timers.update(1.0)
    assert len(ticks) == 4
    assert not timers.active(handle)


def test_every_with_a_count_retires_itself():
    timers = loom.Timers()
    ticks = []
    timers.every(0.1, lambda: ticks.append(1), times=3)

    timers.update(1.0)                # long enough for far more than three
    assert len(ticks) == 3
    assert timers.count == 0


def test_a_timer_lives_without_python_keeping_a_reference():
    # The engine owns the callback: a game writes `timers.after(1, self.spawn)`
    # and keeps nothing. If the C++ side did not hold the callable, this would
    # either not fire or crash.
    timers = loom.Timers()
    calls = []

    def make():
        timers.after(0.5, lambda: calls.append("fired"))

    make()
    timers.update(0.5)
    assert calls == ["fired"]


def test_a_callback_can_schedule_and_cancel_timers():
    timers = loom.Timers()
    order = []
    state = {}

    def second():
        order.append("second")
        timers.cancel(state["repeat"])

    def first():
        order.append("first")
        timers.after(1.0, second)

    state["repeat"] = timers.every(1.0, lambda: order.append("tick"))
    timers.after(1.0, first)

    timers.update(1.0)                # first + the repeat's first tick
    assert order == ["tick", "first"]
    timers.update(1.0)                # the repeat ticks, then second cancels it
    assert order == ["tick", "first", "tick", "second"]
    timers.update(5.0)                # nothing left running
    assert order == ["tick", "first", "tick", "second"]
    assert timers.count == 0


def test_clear_stops_everything():
    timers = loom.Timers()
    calls = []
    timers.after(1.0, lambda: calls.append(1))
    timers.every(1.0, lambda: calls.append(2))
    assert timers.count == 2

    timers.clear()
    assert timers.count == 0
    timers.update(5.0)
    assert calls == []


def test_game_drives_timers():
    game = loom.Game()
    calls = []
    game.timers.after(0.5, lambda: calls.append("boom"))

    # `game.timers` is the engine's own Timers, not a copy: updating it through
    # the property is what the run loop does every frame.
    game.timers.update(0.5)
    assert calls == ["boom"]
    assert game.timers.count == 0

File: tests/cpp/test_timers.cpp
#include <gtest/gtest.h>
#include "core/timers.hpp"

using namespace loom;

TEST(Timers, AfterFiresOnceWhenTheDelayElapses) {
    Timers timers;
    int fired = 0;
    timers.after(1.f, [&] { ++fired; });

    timers.update(0.5f);
    EXPECT_EQ(fired, 0);          // not yet
    timers.update(0.5f);
    EXPECT_EQ(fired, 1);
    timers.update(5.f);
    EXPECT_EQ(fired, 1);          // and never again
    EXPECT_EQ(timers.count(), 0u);
}

TEST(Timers, AfterZeroFiresOnTheNextUpdate) {
    Timers timers;
    int fired = 0;
    timers.after(0.f, [&] { ++fired; });

    EXPECT_EQ(fired, 0);          // scheduling alone does not run it
    timers.update(0.016f);
    EXPECT_EQ(fired, 1);
}

TEST(Timers, EveryRepeatsForever) {
    Timers timers;
    int fired = 0;
    timers.every(0.5f, [&] { ++fired; });

    for (int i = 0; i < 10; ++i) timers.update(0.5f);
    EXPECT_EQ(fired, 10);
    EXPECT_EQ(timers.count(), 1u);
}

TEST(Timers, EveryWithACountRetiresItself) {
    Timers timers;
    int fired = 0;
    timers.every(1.f, [&] { ++fired; }, 3);

    for (int i = 0; i < 10; ++i) timers.update(1.f);
    EXPECT_EQ(fired, 3);
    EXPECT_EQ(timers.count(), 0u);
}

TEST(Timers, RepeatCarriesTheOvershootRatherThanDrifting) {
    Timers timers;
    int fired = 0;
    timers.every(1.f, [&] { ++fired; });

    // 0.6s of overshoot on the first fire must not be thrown away, or a timer
    // driven by uneven frames slowly loses time.
    timers.update(1.6f);
    EXPECT_EQ(fired, 1);
    timers.update(0.4f);          // total 2.0s -> the second fire is due
    EXPECT_EQ(fired, 2);
}

TEST(Timers, ALongFrameCatchesUpOnEveryMissedFire) {
    Timers timers;
    int fired = 0;
    timers.every(0.1f, [&] { ++fired; });

    timers.update(1.f);           // ten intervals in one frame
    EXPECT_EQ(fired, 10);
}

TEST(Timers, EveryZeroFiresOncePerFrameNotForever) {
    Timers timers;
    int fired = 0;
    timers.every(0.f, [&] { ++fired; });   // degenerate: would spin if unguarded

    timers.update(0.016f);
    timers.update(0.016f);
    EXPECT_EQ(fired, 2);
}

TEST(Timers, CancelStopsATimerBeforeItFires) {
    Timers timers;
    int fired = 0;
    TimerHandle h = timers.after(1.f, [&] { ++fired; });
    EXPECT_TRUE(timers.active(h));

    EXPECT_TRUE(timers.cancel(h));
    EXPECT_FALSE(timers.active(h));
    EXPECT_FALSE(timers.cancel(h));        // already gone

    timers.update(2.f);
    EXPECT_EQ(fired, 0);
}

TEST(Timers, CancelIsSafeFromInsideTheCallback) {
    Timers timers;
    int fired = 0;
    TimerHandle h = 0;
    h = timers.every(1.f, [&] {
        ++fired;
        timers.cancel(h);          // a repeat that stops itself on some condition
    });

    timers.update(1.f);
    timers.update(10.f);
    EXPECT_EQ(fired, 1);
    EXPECT_EQ(timers.count(), 0u);
}

TEST(Timers, ACallbackCanScheduleAnotherTimer) {
    Timers timers;
    int first = 0, second = 0;
    timers.after(1.f, [&] {
        ++first;
        timers.after(1.f, [&] { ++second; });
    });

    timers.update(1.f);
    EXPECT_EQ(first, 1);
    EXPECT_EQ(second, 0);         // the new timer starts counting next frame,
                                  // it does not inherit this frame's overshoot
    timers.update(1.f);
    EXPECT_EQ(second, 1);
}

TEST(Timers, ClearFromInsideACallbackStopsTheRest) {
    Timers timers;
    int a = 0, b = 0;
    timers.every(1.f, [&] { ++a; timers.clear(); });
    timers.every(1.f, [&] { ++b; });

    timers.update(1.f);
    EXPECT_EQ(a, 1);
    EXPECT_EQ(b, 0);              // cleared before it got its turn
    EXPECT_EQ(timers.count(), 0u);

    timers.update(1.f);
    EXPECT_EQ(a, 1);
    EXPECT_EQ(b, 0);
}

TEST(Timers, ClearDropsEverything) {
    Timers timers;
    int fired = 0;
    timers.after(1.f, [&] { ++fired; });
    timers.every(1.f, [&] { ++fired; });
    EXPECT_EQ(timers.count(), 2u);

    timers.clear();
    EXPECT_EQ(timers.count(), 0u);
    timers.update(5.f);
    EXPECT_EQ(fired, 0);
}

TEST(Timers, HandlesAreDistinctAndNeverZero) {
    Timers timers;
    TimerHandle a = timers.after(1.f, [] {});
    TimerHandle b = timers.after(1.f, [] {});
    EXPECT_NE(a, 0u);
    EXPECT_NE(b, 0u);
    EXPECT_NE(a, b);

    // An empty callback is not schedulable — nothing to call.
    EXPECT_EQ(timers.after(1.f, nullptr), 0u);
    EXPECT_FALSE(timers.active(0));
}
